### PR TITLE
Add read lock to get_flag_keys method

### PR DESCRIPTION
**Pull request by schmit**
---
labels: mergeable
---
[//]: #  (Link to the issue corresponding to this chunk of work)
Fixes: #__issue__

## Motivation and Context
[//]: #  (Why is this change required? What problem does it solve?)

Adding a read lock to the `get_flag_keys` method to ensure we do not hit a race condition. Also:
- refactored the Read/Write lock class to use a context manager for more readable code
- replace the LRUCache with a simple dictionary

## How has this been tested?
[//]: # (Please describe in detail how you tested your changes)

Tests pass

[//]: # (OPTIONAL)
[//]: # (Add one or multiple labels: enhancement, refactoring, bugfix)

### CODE REVIEW

File: eppo_client/configuration_store.py [hunk 1]
```diff
@@ -1,33 +1,25 @@
...
 from eppo_client.read_write_lock import ReadWriteLock
 
 T = TypeVar("T")
 
 
 class ConfigurationStore(Generic[T]):
-    def __init__(self, max_size: int):
-        self.__cache: LRUCache = LRUCache(maxsize=max_size)
+    def __init__(self):
+        self.__cache: Dict[str, T] = {}
```
Comment (aarsilv) [inline eppo_client/configuration_store.py:10]: How is this cache used? If it's for _assignments_ I think we do want some sort of eviction

**Review by aarsilv — COMMENTED** [review]
Just commenting instead of approving as this has exceed my comfort level reading python 😅

File: eppo_client/configuration_store.py (hunk 1 above)
Comment (schmit) [reply eppo_client/configuration_store.py:10]: This cache is used to store the flag configuration. There is no cache for assignments for the Python SDK at the moment.

File: eppo_client/read_write_lock.py [hunk 2]
```diff
@@ -1,6 +1,7 @@
 import threading
+from contextlib import contextmanager
 
-# Copied from: https://www.oreilly.com/library/view/python-cookbook/0596001673/ch06s04.html
+# Adapted from: https://www.oreilly.com/library/view/python-cookbook/0596001673/ch06s04.html
```
Comment (leoromanovsky) [inline eppo_client/read_write_lock.py:4]: Show me the LLM vectors

File: eppo_client/configuration_store.py [hunk 3]
```diff
@@ -1,33 +1,25 @@
...
-        finally:
-            self.__lock.release_read()
+        with self.__lock.reader():
+            return self.__cache.get(key, None)
 
     def set_configurations(self, configs: Dict[str, T]):
-        try:
-            self.__lock.acquire_write()
+        with self.__lock.writer():
             self.__cache.clear()
```
Comment (leoromanovsky) [inline eppo_client/configuration_store.py:19]: No stale keys here!

File: eppo_client/configuration_store.py [hunk 4]
```diff
@@ -1,33 +1,25 @@
...
 
     def get_configuration(self, key: str) -> Optional[T]:
-        try:
-            self.__lock.acquire_read()
-            return self.__cache[key]
-        except KeyError:
-            return None  # key does not exist
-        finally:
-            self.__lock.release_read()
+        with self.__lock.reader():
```
Comment (leoromanovsky) [inline eppo_client/configuration_store.py:14]: We aren't worried about two readers - I thought the problem is not wanting to read during a write? If so perhaps one lock is sufficient

Comment (schmit) [reply eppo_client/configuration_store.py:14]: We block writing when there are >= 1 readers until they are done.

See implementation [here](https://github.com/Eppo-exp/python-sdk/blob/94b597f29a0d68fd90e91d5759e8b21733878b2d/eppo_client/read_write_lock.py)

File: eppo_client/read_write_lock.py (hunk 2 above)
Comment (schmit) [reply eppo_client/read_write_lock.py:4]: I tried that first, it went down a much more complicated path so I just added the contextmanager myself :)

File: eppo_client/configuration_store.py (hunk 4 above)
Comment (leoromanovsky) [reply eppo_client/configuration_store.py:14]: Nice!!

**Review by leoromanovsky — APPROVED** [review]

**Pull request merged**